extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,762 @@
+        partition.getStream(),
+        partition.getPartitionId(),
+        iteratorEnum.toString(),
+        sequenceNumber
+    ).getShardIterator();
+
+    checkPartitionsStarted = true;
+  }
+
+  private void filterBufferAndResetFetchRunnable(Set<StreamPartition<String>> partitions) throws InterruptedException
+  {
+    scheduledExec.shutdown();
+
+    try {
+      if (!scheduledExec.awaitTermination(EXCEPTION_RETRY_DELAY_MS, TimeUnit.MILLISECONDS)) {
+        scheduledExec.shutdownNow();
+      }
+    }
+    catch (InterruptedException e) {
+      log.warn(e, "InterruptedException while shutting down");
+      throw e;
+    }
+
+    scheduledExec = Executors.newScheduledThreadPool(
+        fetchThreads,
+        Execs.makeThreadFactory("KinesisRecordSupplier-Worker-%d")
+    );
+
+    // filter records in buffer and only retain ones whose partition was not seeked
+    BlockingQueue<OrderedPartitionableRecord<String, String>> newQ = new LinkedBlockingQueue<>(recordBufferSize);
+    records
+        .stream()
+        .filter(x -> !partitions.contains(x.getStreamPartition()))
+        .forEachOrdered(newQ::offer);
+
+    records = newQ;
+
+    // restart fetching threads
+    partitionResources.values().forEach(x -> x.started = false);
+    checkPartitionsStarted = true;
+  }
+
+  @Nullable
+  private String getSequenceNumberInternal(StreamPartition<String> partition, ShardIteratorType iteratorEnum)
+  {
+
+    String shardIterator = null;
+    try {
+      shardIterator = kinesis.getShardIterator(
+          partition.getStream(),
+          partition.getPartitionId(),
+          iteratorEnum.toString()
+      ).getShardIterator();
+    }
+    catch (ResourceNotFoundException e) {
+      log.warn(e, "Caught ResourceNotFoundException while getting shardIterator");
+    }
+
+    return getSequenceNumberInternal(partition, shardIterator);
+  }
+
+  @Nullable
+  private String getSequenceNumberInternal(StreamPartition<String> partition, String shardIterator)
+  {
+    long timeoutMillis = System.currentTimeMillis() + fetchSequenceNumberTimeout;
+
+    while (shardIterator != null && System.currentTimeMillis() < timeoutMillis) {
+
+      if (closed) {
+        log.info("KinesisRecordSupplier closed while fetching sequenceNumber");
+        return null;
+      }
+
+      GetRecordsResult recordsResult;
+      try {
+        recordsResult = kinesis.getRecords(new GetRecordsRequest().withShardIterator(shardIterator).withLimit(1000));
+      }
+      catch (ProvisionedThroughputExceededException e) {
+        try {
+          Thread.sleep(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS);
+          continue;
+        }
+        catch (InterruptedException e1) {
+          log.warn(e1, "Thread interrupted!");
+          Thread.currentThread().interrupt();
+          break;
+        }
+      }
+
+      List<Record> records = recordsResult.getRecords();
+
+      if (!records.isEmpty()) {
+        return records.get(0).getSequenceNumber();
```

Comment:
If you're only ever expecting one record here, can you check and throw an exception if you get more?

Reply:
it is fine if we get more than 1 record, `kinesis.getRecords(new GetRecordsRequest().withShardIterator(shardIterator).withLimit(1000));` fetches up to 1000 records to make sure we can find the first record in this shard (in the case where the shard is constantly removing records due to records being past retention period).

Reply:
Got it, I remember this conversation now.